res/controllers/Numark-Party-Mix.scripts.js

```diff
@@ -0,0 +1,440 @@
+            inKey: "enabled",
+            outKey: "enabled",
+            outConnect: false
+        });
+    })
+    p[3] = new components.Button({
+        midi: [0x93 + deckNumber, 0x17],
+        type: components.Button.prototype.types.toggle,
+        group: "[EffectRack1_EffectUnit" + deckNumber + "]",
+        inKey: "mix_mode",
+        outKey: "mix_mode",
+        output: function() {
+            if (engine.getParameter("[EffectRack1_EffectUnit" + deckNumber + "]", "mix_mode") === 0) {
+                midi.sendShortMsg(0x93 + deckNumber, 0x17, 0x7F); // cue
+            } else {
+                midi.sendShortMsg(0x83 + deckNumber, 0x17, 0x01); // cue
+            }
+        },
+        outConnect: false
+    });
+};
+NumarkPartyMix.ModeEFX.prototype = Object.create(components.ComponentContainer.prototype);
+
+NumarkPartyMix.Browse = function() {
+    var pressturn = false;
+
+    this.knob = new components.Encoder({
+        input: function(channel, control, value) {
+            var direction;
+            if (pressturn) {
+                if (value > 0x40) {
+                    engine.setParameter("[Library]", "MoveFocusForward", 1);
+                } else {
+                    engine.setParameter("[Library]", "MoveFocusBackward", 1);
+                }
+            } else {
+                direction = (value > 0x40) ? -1 : 1;
+                engine.setParameter("[Library]", "MoveVertical", direction);
+            }
+        }
+    });
+
+    this.knobButton = new components.Button({
+        group: "[Library]",
+        type: components.Button.prototype.types.powerWindow,
+        isLongPressed: false,
+        input: function(channel, control, value, status, group) {
+            pressturn = this.isPress(channel, control, value, status);
+            print("pressturn");
+            print(pressturn);
+            if (pressturn) {
+                this.inToggle(channel, control, value, status, group);
+                this.isLongPressed = false;
+                this.longPressTimer = engine.beginTimer(this.longPressTimeout, function() {
+                    this.isLongPressed = true;
+                    this.longPressTimer = 0;
+                }, true);
+            } else {
+                this.inToggle(channel, control, value, status, group);
+                if (this.longPressTimer !== 0) {
+                    engine.stopTimer(this.longPressTimer);
+                    this.longPressTimer = 0;
+                }
+                this.isLongPressed = false;
+            }
+        },
+        inToggle: function(channel, control, value, status, group) {
+            if (! (pressturn || this.isLongPressed)) {
+                engine.setParameter("[Library]", "GoToItem", 1);
+            }
+        }
+    });
+};
+NumarkPartyMix.Browse.prototype = new components.ComponentContainer();
+
+NumarkPartyMix.Gains = function() {
+    this.mainGain = new components.Pot({
+        group: "[Master]",
+        inKey: "gain"
```

Comment:
Usually this control should not be mapped. It is very sensitive to clipping, and it can easily f**k up the main mix if users don't know the difference between gain and volume. See this for an explanation of gain staging https://manual.mixxx.org/2.3/en/chapters/user_interface.html#headphone-and-main-mix-controls

If this knob affects the hardware output level it **must** not be mapped!

Reply:
previous versions of this controller also mapped the master. https://github.com/fiveangle/mixxx_numark_partymix/blob/dfd5b6787ba6daa5f1cc1b5c22e9a65eb96256f9/Numark-Party-Mix.midi.xml#L56 buildin hardware doesn't has a alsa volume control.

Reply:
yes it only controls the volume in mixxx the buildin dac doesn't have any volume control.

Reply:
Thanks for clarifying.